api

```diff
@@ -675,6 +675,73 @@ debian_ppa_installer() { #setup a PPA on a Debian distro. Arguments: ppa_name di
+  elif [ "$1" == 'refresh' ];then
+    action=refresh
+  else
+    error "terminal_manage(): Must specify an action: either 'install' or 'uninstall' or 'update' or 'refresh'"
+  fi
+  
+  app="$2" #one app name (or filelist of the format 'filelist:path/to/file:path/to/another/file')
+  terminal_manage_multi "$action $app"
+}
+
+terminal_manage_multi() { #function to install/uninstall/update/refresh multiple apps (or filelists) - uses a terminal and refreshes the app list
+  queue="$1" #one or multiple actions and app names (or filelists of the format 'filelist:path/to/file:path/to/another/file')
+  
+  #To prevent multiple simultaneous manage instances, use the 'daemon' mode. This will create a queue of actions that are executed concurrently.
+  #The first daemon instance is the 'master' process. Subsequent processes will add the action to the queue and then exit.
+  #All output is generated by the 'master' daemon process. Subsequent processes shouldn't open a terminal, because the master one is already open.
+  if [ -f "${DIRECTORY}/data/manage-daemon/pid" ] && [ -f "/proc/$(cat "${DIRECTORY}/data/manage-daemon/pid")/status" ];then
+    #The 'master' daemon is already running. Avoid launching a second terminal.
+    "${DIRECTORY}/manage" daemon "$queue"
+    
+  else
+    #in a terminal, first get the api functions, display the pi-apps logo, run the manage script, and refresh the app list if the $pipe variable is set
+    "${DIRECTORY}/etc/terminal-run" '
+      DIRECTORY="'"$DIRECTORY"'"
+      export geometry2="'"$geometry2"'"
+      source "${DIRECTORY}/api"
+      generate_logo
+      
+      refresh_list() { #Refresh the current list of apps in the event of a change
+        if [ ! -z "'"$pipe"'" ];then
+          echo -e "\f" > "'"$pipe"'"
+          "${DIRECTORY}/preload" yad "'"$prefix"'" > "'"$pipe"'" 2>/dev/null
+        fi
+      }
+      
+      "${DIRECTORY}/manage" daemon "'"$queue"'"
+      
+      #refresh app list
+      refresh_list
+      
+      for i in {30..1} ;do
+        echo -en "You can close this window now. Auto-closing in $i seconds.\e[0K\r"
+        sleep 1
+      done
+      ' "Terminal Output"
+    
+    # Check if terminal-run failed to launch. GUI users don't see any terminal output if it fails (since there is no terminal open) so we need to prompt them with a GUI window
+    if [ "$?" != 0 ]; then
+      echo -e "Unable to open a terminal.\nDebug output below.\n$(DEBUG=1 "${DIRECTORY}/etc/terminal-run" 2>&1)" | yad --center --window-icon="${DIRECTORY}/icons/logo.png" \
+      --width=700 --height=300 --text-info --title="Error occured when calling terminal-run" \
+      --image="${DIRECTORY}/icons/error.png" --image-on-top --fontname=12 \
+      --button='OK'
+      if echo "$queue" | grep -q "^update filelist:" ;then #list of files separated by :
+        updatable_apps='' updatable_files="$(echo "$queue" | grep "^update filelist:" | sed 's/^update filelist://g' | tr ':' '\n')" no_status=true update_now_cli
```

Comment:
This is an api function that calls `update_now_cli`, a function that I am pretty sure is still in the `manage` script, though I have not finished reviewing your changes. Did you assume that `manage` is already sourced when it reaches this point in this terminal_manage_multi function?

Reply:
`update_now_cli` has not changed (except fixing the error code) and still resides in the `updater` script. `terminal_manage_multi` is being called from the `updater` so the function is guaranteed to be available